openmc/lattice.py

```diff
@@ -759,6 +770,277 @@ def is_valid_index(self, idx):
+                find_lattice_neighbors(i, j)
+
+                pattern[1, 1] = getattr(self.universes[j][i], attribute)
+
+                # Look for pattern in dictionary of patterns found
+                found = False
+                for known_pattern, pattern_data in patterns.items():
+
+                    # Degenerate discretizations use unique patterns
+                    if strategy == "degenerate":
+                        break
+
+                    # Look at all rotations of pattern
+                    for rot in range(4):
+                        if not found and tuple(map(tuple, pattern)) == known_pattern:
+                            found = True
+
+                            # Save location of the pattern in the lattice
+                            pattern_data['locations'].append((i, j))
+
+                        # Rotate pattern
+                        pattern = np.rot90(pattern)
+
+                    # Look at transpose of pattern and its rotations
+                    pattern = np.transpose(pattern)
+                    for rot in range(4):
+                        if not found and \
+                             tuple(map(tuple, pattern)) == known_pattern:
+                            found = True
+
+                            # Save location of the pattern in the lattice
+                            pattern_data['locations'].append((i, j))
+
+                        # Rotate pattern
+                        pattern = np.rot90(pattern)
+
+                    # Transpose pattern back for the next search
+                    pattern = np.transpose(pattern)
+
+                # Create new pattern and add to the patterns dictionary
+                if not found:
+                    pattern_data = {}
+                    pattern_data['locations'] = [(i, j)]
+                    if strategy == "lns":
+                        patterns[tuple(map(tuple, pattern))] = pattern_data
+                    elif strategy == "degenerate":
+                        patterns[(i,j)] = pattern_data
+
+        # Discretize lattice
+        for pattern, pattern_data in patterns.items():
+
+            first_pos = pattern_data['locations'][0]
+
+            # Create a clone of the universe, without cloning materials
+            new_universe = self.universes[first_pos[1]][first_pos[0]].clone(
+                 clone_materials=False, clone_regions=False)
+
+            # Discretize the immediate sub-lattices of this universe
+            for sub_cell in new_universe.cells.values():
+
+                sub_universe = sub_cell.fill
+
+                try:
+                    sub_universe.discretize(strategy,
+                                            universes_to_ignore,
+                                            materials_to_clone,
+                                            lattice_neighbors=[],
+                                            attribute=attribute)
+                except AttributeError:
```

Comment:
I know I suggested `AttributeError`, but a whole lot of things can throw this error so its better to directly check `if hasattr(sub_universe, discretize)` instead of using a `try`

Reply:
I'm getting rid of the recursive call. I think in a lot of cases, the 'key' function will only be valid at a certain lattice level, therefore passing it down with recursion wont work